Add formula highlighting in Editor

ax2mx opened the pull request:
- Handle correct caret position on formula modification
- Add «black-box» tokenizer function
- Add unparsed state, input & paste handling
- Move autoSizeEditor() into CellEditor.svelte
- Change editor click logic
- Fix toggle comment support
- Add search feature in Node Navigator

ax2mx committed:
Add initial formula highlighting

ax2mx committed:
Enhance formula highlight logic

- Enhance highlight and spanify logic
- DRY highlight()
- Add «black box» getTokens() function
- Add unparsed state, input & paste handling
- Move autoSizeEditor() to CellEditor.svelte

ax2mx committed:
Fix wrong caret position on formula input

ax2mx committed:
Fix editor click logic

ax2mx committed:
Add formula highlight support

- Add node insertions
- Change editor click logic
- Fix caret position bug on formula input
- Fix toggle comment in editor mode

dmaevsky commented:
![](https://github.trello.services/images/mini-trello-icon.png) [asdadsfg](https://trello.com/c/PgmvmI9Z/45-asdadsfg)

ax2mx committed:
Remove comments

ax2mx committed:
Fix empty input handling

ax2mx committed:
Various fixes

ax2mx committed:
Add search feature in Node Navigator

- Add search feature
- Remove @apply statements
- Fix layout and styling

ax2mx committed:
Merge branch 'master' into master

ax2mx committed:
Merge branch 'master' into master

ax2mx committed:
Merge branch 'dmaevsky:master' into master

ax2mx committed:
Add case insensitive search

ax2mx committed:
Close editor when navigating nodes

ax2mx committed:
Fix highlight z-index above Node Navigator

ax2mx committed:
DRY panel toggling function

ax2mx committed:
Fix insert in partly parsed formula

dmaevsky commented on src/sandbox/components/CellEditor.svelte:
```diff
@@ -1,23 +1,228 @@
+      parser.parse(str);
+      return parser.nodes;
+    } catch (e) {
+      console.log("Formula parsing error: ", e);
```
console.debug ? or remove

ax2mx replied:
I think it might be better to leave this notification at debug level.
Fixed.

dmaevsky commented on src/sandbox/components/CellEditor.svelte:
```diff
@@ -1,23 +1,228 @@
+  let isPartlyParsed = false;
+  let innerHTML;
+
+  const parser = new ProgressiveEval(environment, () => {});
```
better initialize it directly in getTokens

ax2mx replied:
Fixed.

dmaevsky commented on src/sandbox/components/CellEditor.svelte:
```diff
@@ -1,23 +1,228 @@
+
+    const spacer = formula.substring(0, formula.length - rightHand.length - 1);
+
+    let result = `<span id="highlight" class="highlighted_string">`
```
"#highlight" is too common as id, may have an id like this in the app itself

dmaevsky commented on src/sandbox/components/CellEditor.svelte:
```diff
@@ -1,23 +1,228 @@
+      }
+      if (found) break;
+      caretPosition += nodes[i].textContent.length;
+    }
```
I believe you, but this is SO NOT CLEAR. Need comments

dmaevsky commented on src/sandbox/components/CellEditor.svelte:
```diff
@@ -1,23 +1,228 @@
+  on:paste={(e) => {
+    e.preventDefault();
+    let text = (e.originalEvent || e).clipboardData.getData('text/plain');
+    document.execCommand("insertHTML", false, text);
```
https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand
Deprecated: This feature is no longer recommended. 
No other option here?

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -142,11 +146,51 @@
     highlight = [row, col];
   }
 
+  function getPosition(anchor, offset) {
+
+    const highlight = document.querySelector("#highlight");
+
+    if (!highlight) return offset;
+
+    const anchorParent = anchor.parentNode;
+    const nodes = highlight.childNodes;
+
+    let caretPosition = 0;
+    let found = false;
+
+    for (let i = 0; i < nodes.length; i++) {
+      const children = nodes[i].childNodes;
+
+      if (nodes[i] === anchorParent || nodes[i] === anchor) {
+        found = true;
+        caretPosition += offset;
+        break;
+      } else if (children.length > 1) {
+        for (let j = 0; j < children.length; j++) {
+          caretPosition += children[j].textContent.length;
+          if (children[j] === anchorParent) {
+            found = true;
+            break;
+          }
+        }
+      }
+      if (found) break;
+      caretPosition += nodes[i].textContent.length;
+    }
+
+    if (!found) caretPosition = editorSession.length;
+
+    return caretPosition;
+  }
+
```
I just saw the same code in CellEditor.svelte ?

dmaevsky commented on src/sandbox/components/NodeNavigator.svelte:
```diff
@@ -36,67 +37,63 @@
-
-  .node:hover .id {
-    opacity: 1;
+  .hidden {
+    display: none;
   }
```
This is a standard TW class, normally not needed ?

ax2mx committed:
PR review fixes

dmaevsky commented on src/utils/ui.js:
```diff
@@ -0,0 +1,3 @@
+export function togglePanel(selector) {
```
may be just name it "toggle"?

ax2mx committed:
Add ellx- prefix to elements ids

ax2mx committed:
Rename onInput flag to isChangedByInput

ax2mx committed:
remoe unused element class

dmaevsky commented on src/sandbox/components/CellEditor.svelte:
```diff
@@ -1,23 +1,197 @@
+
+  import ProgressiveEval from "../../runtime/engine/progressive_assembly.js";
+  import * as environment from "../../runtime/engine/reserved_words.js";
+  import { getCaretPosition } from "../../utils/highlight";
```
extension .js 
It works with our bundler, but maybe not forever. At least this is the convention for now: we don't rely on node resolution algorithm whenever possible

dmaevsky commented on src/sandbox/components/Grid.svelte:
```diff
@@ -32,6 +35,8 @@
   let highlight = selection;
   let arrowRow, arrowCol;
 
+  $: if (selection && !document.querySelector("#ellx-node-navigator").classList.contains("hidden")) closeEditor();
```
better to make NodeNavigator hidden flag part of the store. Let's discuss: this is a very good example when state management becomes necessary

ax2mx committed:
Fix editor id typo

ax2mx committed:
Handle Node Navigator visibility as app state

ax2mx committed:
Simplify highlightInput logic

- Make autoSizeEditor() sync

ax2mx committed:
Handle Shortcut Helper visibility as app state

ax2mx committed:
Drastically simplify highlight logic

- Unify on mount, on insert and on input cases
- Get rid of handleInput()
- Get rid of redundant flag isChangedByInput
- Fix setCaretPosition logic
- Drop caretPosition to undefined on closing editor

ax2mx committed:
Get rid of togglePanel

ax2mx committed:
Merge remote-tracking branch 'upstream/master'
